File: server/src/models/pod.rs
use std::collections::HashMap;

use diesel;
use diesel::prelude::*;
use uuid::Uuid;
use chrono::{DateTime, Utc, Duration};

use data::types::*;
use data::modules::get_module_list;
use helpers::db::DB;

use models::module::Module;
use models::resource::Resource;
use models::research::Research;

use schema::pods;
use schema::modules::dsl as module_dsl;
use schema::resources::dsl as resource_dsl;
use schema::researches::dsl as research_dsl;


#[derive(Debug, Serialize, Deserialize, Identifiable, Queryable, Associations, AsChangeset)]
#[belongs_to(users)]
#[belongs_to(bases)]
pub struct Pod {
    pub name: String,
    pub id: Uuid,
    pub user_id: Uuid,
    pub base_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}


impl Pod {
    pub fn new_pod(name: String, user_id: Uuid, db: &DB) -> Self {
        // New pod
        let new_pod = NewPod {
            name: format!("{}'s Pod", name),
            user_id: user_id,
        };

        diesel::insert(&new_pod)
            .into(pods::table)
            .get_result::<Pod>(&**db)
            .expect("Error inserting new pod into database.")
    }

    pub fn get_module(&self, id: Uuid, db: &DB) -> Result<Module, diesel::result::Error> {
        // Get the research from an id
        module_dsl::modules
            .filter(module_dsl::id.eq(id))
            .filter(module_dsl::pod_id.eq(self.id))
            .get_result::<Module>(&**db)
    }

    pub fn get_modules(&self, db: &DB) -> Result<Vec<Module>, diesel::result::Error> {
        module_dsl::modules
            .filter(module_dsl::pod_id.eq(self.id))
            .get_results::<Module>(&**db)
    }

    pub fn get_researches(&self, db: &DB) -> Vec<Research> {
        research_dsl::researches
            .filter(research_dsl::pod_id.eq(self.id))
            .get_results::<Research>(&**db)
            .unwrap()
    }

    pub fn get_research(&self, name: String, db: &DB) -> Result<Research, diesel::result::Error> {
        // Get the research from an id
        research_dsl::researches
            .filter(research_dsl::pod_id.eq(self.id))
            .filter(research_dsl::name.eq(name))
            .get_result::<Research>(&**db)
    }

    pub fn get_resources(&self, db: &DB) -> Vec<Resource> {
        resource_dsl::resources
            .filter(resource_dsl::pod_id.eq(self.id))
            .get_results(&**db)
            .expect("Failed to get user resources.")
    }

    pub fn has_enough_resources(&self, costs: &Option<Vec<(ResourceTypes, i64)>>, db: &DB) -> bool {
        let pod_resources = self.get_resources(db);
        Resource::enough_resources(costs, pod_resources, db)
    }

    pub fn update_resource_production(&self, db: &DB) {
        // Create a map to save the resource production.
        let mut resources_production: HashMap<ResourceTypes, i64> = HashMap::new();
        // Get all resources and all modules
        let resources = self.get_resources(db);
        let modules_result = self.get_modules(db);
        if let Ok(modules) = modules_result {
            let module_list = get_module_list();
            for module in modules {
                // Get all modules information to compute the current resource production
                let module_type =
                    ModuleTypes::from_string(&module.name).expect("Missing module type");
                let level = &module_list
                    .get(&module_type)
                    .as_ref()
                    .expect("No module in yml for this type.")
                    .levels
                    [module.level as usize];

                // Calculate production
                for &(ref resource_ref, amount) in &level.generates {
                    let resource_type = ResourceTypes::from_string(&resource_ref.to_string())
                        .unwrap();
                    *resources_production.entry(resource_type).or_insert(0) += amount;
                }

                // Calculate consumption
                for &(ref resource_ref, amount) in &level.consumes {
                    let resource_type = ResourceTypes::from_string(&resource_ref.to_string())
                        .unwrap();
                    *resources_production.entry(resource_type).or_insert(0) -= amount;
                }
            }

            for mut resource in resources {
                let resource_type = ResourceTypes::from_string(&resource.name).unwrap();
                match resources_production.get(&resource_type) {
                    // Calculate produced amount.
                    Some(amount) => {
                        let elapsed_time: Duration =
                            Utc::now().signed_duration_since(resource.updated_at);
                        let produced_since_last_update: i64 =
                            (resource.production * elapsed_time.num_milliseconds()) / 60 / 60 /
                                1000;
                        resource.change_resource(produced_since_last_update, false, db);
                        diesel::update(resource_dsl::resources.filter(
                            resource_dsl::id.eq(resource.id),
                        )).set(resource_dsl::production.eq(amount))
                            .execute(&**db)
                            .expect("Failed to update resource production.");
                    }
                    None => (),
                }
            }
        }
    }

    pub fn update_resources(&self, db: &DB) {
        let resources = self.get_resources(db);
        for mut resource in resources {
            let elapsed_time: Duration = Utc::now().signed_duration_since(resource.updated_at);
            let produced_since_last_update: i64 =
                (resource.production * elapsed_time.num_milliseconds()) / 60 / 60 / 1000;
            resource.change_resource(produced_since_last_update, false, db);
        }
    }
}

#[derive(Insertable)]
#[table_name = "pods"]
pub struct NewPod {
    pub name: String,
    pub user_id: Uuid,
}

File: server/models/pod.py
"""Pod database model."""

import uuid
from datetime import datetime, timedelta
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy import (
    func,
    Column,
    ForeignKey,
)

from sqlalchemy.types import (
    String,
    DateTime,
)

from server.extensions import db
from server.models.queue import Queue
from server.models.resource import Resource
from server.data.types import ResourceTypes


class Pod(db.Model):
    """Pod model."""

    __tablename__ = 'pod'

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name = Column(String(255), nullable=False)
    user_id = Column(UUID(as_uuid=True), ForeignKey('user.id', deferrable=True, initially='DEFERRED'), index=True, nullable=False)
    base_id = Column(UUID(as_uuid=True), ForeignKey('base.id'), index=True)

    user = relationship("User", back_populates="pod")

    resources = relationship("Resource")
    queue = relationship("Queue", uselist=False, back_populates="pod")
    modules = relationship("Module", back_populates="pod")
    researches = relationship("Research", back_populates="pod")

    created_at = Column(DateTime, server_default=func.now(), nullable=False)
    updated_at = Column(
        DateTime, server_default=func.now(),
        onupdate=func.current_timestamp(),
        nullable=False,
    )

    def __init__(self, nickname):
        """Create a new pod."""
        self.name = "{nickname}'s pod"
        self.queue = Queue()
        resources = []
        for resource in ResourceTypes:
            resources.append(Resource(resource.name))
        self.resources = resources

    def update_resource_production(self):
        """Update the production of each resource."""
        return

    def update_resources(self):
        """Update the current resource state in the db."""
        for resource in self.resources:
            time_diff = datetime.now() - resource.last_update
            diff_in_seconds = time_diff.total_seconds()
            resource_diff = diff_in_seconds * resource.production / 3600

            resource.amount += resource_diff
            if resource.amount < 0:
                resource.amount = 0
            elif resource.amount > resource.max_amount:
                resource.amount = resource.max_amount
            resource.last_update = datetime.now()

            db.session.add(resource)

    def get_by_name(self, name):
        """Get a resource by name."""
        resource = next((r for r in self.resources if r.name == name), None)
        if resource is None:
            raise LookupError(f'No resource with name {name} found.')
        return resource

    def enough_resources(self, requirements) -> bool:
        """Check if there are enough resources for construction."""
        enough = True
        missing = {}
        self.update_resources()
        for requirement in requirements:
            resource = self.get_by_name(requirement['type'])
            amount = requirement['amount']
            if resource is None:
                print(f'Missing resource: {requirement["type"]}')
            if resource.amount <= amount:
                enough = False
                missing[resource.name] = amount - resource.amount
        return enough, missing

    def subtract_resources(self, requirements) -> bool:
        """Check if there are enough resources for construction."""
        for requirement in requirements:
            resource = self.get_by_name(requirement['type'])
            amount = requirement['amount']
            if (resource.amount - amount) <= 0:
                print(f"Can't afford resource {requirement['type']}: {resource.amount} of {amount}")
                raise Exception
            else:
                resource.amount -= amount
            db.session.add(resource)
        db.session.commit()
        return True

    def add_resources(self, requirements) -> bool:
        """Check if there are enough resources for construction."""
        for requirement in requirements:
            resource = self.get_by_name(requirement['type'])
            amount = requirement['amount']
            resource.amount += amount
            db.session.add(resource)
        db.session.commit()
        return True
